File: src/main/webapp/src/app/policies/policy.ts
import {Customer} from '../customers/customer';
import {Product} from '../products/product';
import {User} from '../users/user';
import {ProductVariant} from '../products/product-variant';
import {DateModel} from 'ng2-datepicker';
import { ProductCategoryAttributeValue } from '../product-categories/product-category';

export const POLICY_STATUS_DRAFT = "DRAFT";
export const POLICY_STATUS_CLOSED = "CLOSED";

interface PolicyAttribute {
  name: string;
  label: string;
  amount: number;
  shouldChangePrice: boolean
}

export class Policy {

  id: number;
  filename: string;
  hasFile: boolean;
  name: string;
  number: string;
  description: string;
  subject: string;
  subjectInfo: string;
  subjectDetails: string;
  status: string = POLICY_STATUS_DRAFT;

  customer: Customer = null;
  product: Product = null;
  user: User = null;

  productVariant: ProductVariant = null;

  price: number = 0;

  createdAt: Date;
  updatedAt: Date;
  closedAt: Date;

  startsAt: Date;
  startsAtPicker: DateModel;
  endsAt: Date;
  endsAtPicker: DateModel;

  assignedAttributes: PolicyAttribute[] = [];

  constructor(data) {
    Object.assign(this, data);

    this.customer = data.customer ? new Customer(data.customer) : null;
    this.product = data.product ? new Product(data.product) : null;
    this.user = data.user ? new User(data.user) : null;

    this.productVariant = data.productVariant ? new ProductVariant(data.productVariant) : null;
  }

  prepareToSave() {
    if (this.startsAtPicker) {
      this.startsAt = this.startsAtPicker.momentObj.toDate();
    }

    if (this.endsAtPicker) {
      this.endsAt = this.endsAtPicker.momentObj.toDate();
    }
  }

  statusDisplay() {

    switch (this.status) {
      case POLICY_STATUS_CLOSED:
        return "Aktywna";

      case POLICY_STATUS_DRAFT:
      default:
        return "Szkic";
    }
  }

  statusLabel() {
    switch (this.status) {
      case POLICY_STATUS_CLOSED:
        return "closed";

      case POLICY_STATUS_DRAFT:
      default:
        return "draft";
    }
  }

  isClosed() {
    return this.status == POLICY_STATUS_CLOSED;
  }

  calculatePrice() {
    this.price = this.productVariant.price;

    if (this.assignedAttributes.length) {
      this.assignedAttributes.forEach(value => {

        if (false == value.shouldChangePrice) {
          return;
        }

        this.price = Math.round(this.price * (value.amount / 100) * 100) / 100;
      });
    }
  }
}

File: src/main/java/agency/entity/Policy.java
package agency.entity;

import lombok.Getter;
import lombok.Setter;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.config.EnableMongoAuditing;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;
import javax.validation.constraints.Future;
import javax.validation.constraints.NotNull;
import java.io.File;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;

enum PolicyStatus {
    DRAFT,
    CLOSED
}

class PolicyAttribute {

    @NotNull @Getter @Setter
    private String name;

    @NotNull @Getter @Setter
    private String label;

    @NotNull @Getter @Setter
    private double amount;

    @NotNull @Getter @Setter
    private boolean shouldChangePrice;
}

@Document(collection = "policies")
@EnableMongoAuditing
public class Policy {

    @Id @Getter
    private String id;

    @Setter @Getter
    private String filename;

    @Getter
    private Boolean hasFile = false;

    @NotNull @Setter @Getter
    private String name;

    @NotNull @Setter @Getter
    private String subject;

    @Setter @Getter
    private String subjectInfo;

    @Setter @Getter
    private String subjectDetails;

    @Setter @Getter
    private String description;

    @DBRef @Setter @Getter
    private User user;

    @DBRef @Setter @Getter @NotNull
    private Customer customer;

    @DBRef @Setter @Getter @NotNull
    private Product product;

    @Setter @Getter @NotNull
    private ProductVariant productVariant;

    @Setter @Getter
    private ArrayList<PolicyAttribute> assignedAttributes = new ArrayList<>();

    @NotNull @Setter @Getter
    private double price;

    @Setter @Getter @NotNull
    private PolicyStatus status = PolicyStatus.DRAFT;

    @CreatedDate
    @Setter @Getter
    private Date createdAt;

    @LastModifiedDate
    @Setter @Getter
    private Date updatedAt;

    @Setter @Getter
    private Date startsAt;

    @Setter @Getter @Future
    private Date endsAt;

    @Setter @Getter
    private Date closedAt;

    public Policy markAsClosed() {
        this.setClosedAt(new Date());
        this.setStatus(PolicyStatus.CLOSED);

        return this;
    }

    public boolean isDraft() {
        return this.status == PolicyStatus.DRAFT;
    }

    public boolean isClosed() {
        return this.status == PolicyStatus.CLOSED;
    }

    public String getStartAtFormated() {
        return getFormatedDate(this.startsAt);
    }

    public String getClosedAtFormated() {
        return getFormatedDate(this.closedAt);
    }

    public String getEndsAtFormated() {
        return getFormatedDate(this.endsAt);
    }

    String getFormatedDate(Date date) {

        if (null == date) {
            return "";
        }

        SimpleDateFormat format = new SimpleDateFormat("dd/MM/yyyy");

        return format.format(date);
    }

    public boolean hasGeneratedFile() {

        File f = new File("storage/policies/" + this.filename);

        if(f.exists() && !f.isDirectory()) {
            return true;
        }

        return false;
    }

    public void checkGeneratedFile() {
        this.hasFile = this.hasGeneratedFile();
    }
}
